primitives/runtime/src/lib.rs

```diff
@@ -468,7 +472,7 @@ pub struct ModuleError {
 	/// Module index, matching the metadata module index.
 	pub index: u8,
 	/// Module specific error value.
-	pub error: u8,
+	pub error: [u8; MAX_MODULE_ERROR_ENCODED_SIZE],
```

Comment:
I have difficulty to find what does this change break.

This is used in the runtime API for block builder: `apply_extrinsic` in `ApplyExtrinsicResult` and some RPC.
I'm not sure how to do the migration, like both client and runtime must upgrade at the same time otherwise ModuleError are not decodable in the client.
I thought the result of the extrinsics were also stored in the client database, but I don't find where. If it is the case we should ensure that old stored value can still be decoded.

maybe a thought would be to add a new variant to DispatchError: `ModuleErrorV2` which olds the new module error with error on 4 bytes ?
This way maybe we could avoid breaking clients ?
Or maybe migration is not as difficult as I think.

Reply:
That is a really really good point!

> I thought the result of the extrinsics were also stored in the client database, but I don't find where. If it is the case we should ensure that old stored value can still be decoded.

That is not done. Nevertheless we can not just break it. Otherwise it would for example mean that you could send a transaction that returns an error, but this results then fails to be decoded on the client side. This means that the validator would discard the entire transaction and you would not have paid for it. This is bad.

To properly fix this, we need to bump the `BlockBuilder` runtime api and version the `apply_extrinsic` runtime api.

Reply:
This should be done in https://github.com/paritytech/substrate/pull/10242/commits/811e2f2edd11167d3fa93b7d05c84454046ea448, please let me know if there's anything else I need to add.